File: dart/editor/tools/plugins/com.google.dart.engine_test/src/com/google/dart/engine/internal/index/file/MemoryNodeManager.java
package com.google.dart.engine.internal.index.file;

import com.google.common.collect.Maps;
import com.google.dart.engine.context.AnalysisContext;

import java.util.List;
import java.util.Map;

/**
 * A {@link NodeManager} that keeps {@link IndexNode}s in memory.
 */
public class MemoryNodeManager implements NodeManager {
  public final StringCodec stringCodec = new StringCodec();
  public final ElementCodec elementCodec = new ElementCodec(stringCodec);
  public final RelationshipCodec relationshipCodec = new RelationshipCodec(stringCodec);

  private final Map<String, IndexNode> nodes = Maps.newHashMap();
  private final Map<String, Integer> nodeLocationCounts = Maps.newHashMap();
  private int locationCount = 0;

  @Override
  public void clear() {
    nodes.clear();
  }

  @Override
  public int getLocationCount() {
    return locationCount;
  }

  @Override
  public IndexNode getNode(String name) {
    return nodes.get(name);
  }

  @Override
  public StringCodec getStringCodec() {
    return stringCodec;
  }

  public boolean isEmpty() {
    for (IndexNode node : nodes.values()) {
      Map<RelationKeyData, List<LocationData>> relations = node.getRelations();
      if (!relations.isEmpty()) {
        return false;
      }
    }
    return true;
  }

  @Override
  public IndexNode newNode(AnalysisContext context) {
    return new IndexNode(context, elementCodec, relationshipCodec);
  }

  @Override
  public void putNode(String name, IndexNode node) {
    // update location count
    {
      locationCount -= getLocationCount(name);
      int nodeLocationCount = node.getLocationCount();
      nodeLocationCounts.put(name, nodeLocationCount);
      locationCount += nodeLocationCount;
    }
    // remember the node
    nodes.put(name, node);
  }

  @Override
  public void removeNode(String name) {
    nodes.remove(name);
  }

  private int getLocationCount(String name) {
    Integer locationCount = nodeLocationCounts.get(name);
    return locationCount != null ? locationCount : 0;
  }
}

File: dart/pkg/analysis_server/lib/src/services/index/store/memory_node_manager.dart
// Copyright (c) 2014, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

library services.src.index.store.store_memory_node_manager;

import 'dart:async';
import 'dart:collection';

import 'package:analysis_server/src/services/index/store/codec.dart';
import 'package:analysis_server/src/services/index/store/split_store.dart';
import 'package:analyzer/src/generated/engine.dart';

class MemoryNodeManager implements NodeManager {
  @override
  StringCodec stringCodec = new StringCodec();

  @override
  ContextCodec contextCodec = new ContextCodec();

  @override
  ElementCodec elementCodec;

  RelationshipCodec _relationshipCodec;

  int _locationCount = 0;
  final Map<String, int> _nodeLocationCounts = new HashMap<String, int>();
  final Map<String, IndexNode> _nodes = new HashMap<String, IndexNode>();

  MemoryNodeManager() {
    elementCodec = new ElementCodec(stringCodec);
    _relationshipCodec = new RelationshipCodec(stringCodec);
  }

  @override
  int get locationCount {
    return _locationCount;
  }

  @override
  void clear() {
    _nodes.clear();
  }

  int getLocationCount(String name) {
    int locationCount = _nodeLocationCounts[name];
    return locationCount != null ? locationCount : 0;
  }

  @override
  Future<IndexNode> getNode(String name) {
    return new Future.value(_nodes[name]);
  }

  bool isEmpty() {
    for (IndexNode node in _nodes.values) {
      Map<RelationKeyData, List<LocationData>> relations = node.relations;
      if (!relations.isEmpty) {
        return false;
      }
    }
    return true;
  }

  @override
  IndexNode newNode(AnalysisContext context) {
    return new IndexNode(context, elementCodec, _relationshipCodec);
  }

  @override
  void putNode(String name, IndexNode node) {
    // update location count
    {
      _locationCount -= getLocationCount(name);
      int nodeLocationCount = node.locationCount;
      _nodeLocationCounts[name] = nodeLocationCount;
      _locationCount += nodeLocationCount;
    }
    // remember the node
    _nodes[name] = node;
  }

  @override
  void removeNode(String name) {
    _nodes.remove(name);
  }
}
